clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/deploy/AWSServerGroupNameResolver.groovy

```diff
@@ -45,6 +52,55 @@ class AWSServerGroupNameResolver extends AbstractServerGroupNameResolver {
 
   @Override
   List<AbstractServerGroupNameResolver.TakenSlot> getTakenSlots(String clusterName) {
-    return asgService.getTakenSlots(clusterName)
+    def cluster = awsCluster(clusterProviders, accountName, clusterName)
+    if (!cluster) {
+      return []
+    }
+
+    def serverGroupsInRegion = cluster.serverGroups.findAll { it.region == region }
+    return serverGroupsInRegion.collect {
+      new AbstractServerGroupNameResolver.TakenSlot(
+        serverGroupName: it.name,
+        sequence       : Names.parseName(it.name).sequence,
+        createdTime    : new Date(it.createdTime)
+      )
+    }
+  }
+
+  @Override
+  String resolveNextServerGroupName(String application, String stack, String details, Boolean ignoreSequence) {
+    def originalNextServerGroupName = super.resolveNextServerGroupName(application, stack, details, ignoreSequence)
+    def nextServerGroupName = originalNextServerGroupName
+    def clusterName = combineAppStackDetail(application, stack, details)
+
+    if (nextServerGroupName) {
+      while (true) {
+        // this resolver uses cached data to determine the next server group name so we should verify it does not already
+        // exist before blindly using it
+        def nextServerGroup = asgService.getAutoScalingGroup(nextServerGroupName)
```

Comment:
Wonder if it might be simpler to do a force cache refresh of the identified 'next' server group and then call the original routine again (and do that until the same result is returned twice, or we circle around, or something like that...). Feels like that would similarly address the race window, but haven't thought too deeply about it.